popde/adaptive_kde.py

```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
         self.set_alpha(new_alpha)
 
 
+class KDEOptimization(AdaptiveBwKDE):
+    """
+    Given grid of bandwidth and alpha values 
+    find optimized values using cross 
+    valitation with figure of merit
+    based on loglikelihood criteria 
+    """
+    def __init__(self, data, weights, bandwidth_options, alpha_options , input_transf=None, stdize=False, rescale=None, backend='KDEpy', bandwidth=0.5, alpha=0.0, dim_names=None, do_fit=True):
+        self.alpha_options = alpha_options
+        self.bandwidth_options = bandwidth_options
+        super().__init__(data, weights, input_transf, stdize,
+                         rescale, backend, bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, bandwidth_val, alpha_val):
+        loo = LeaveOneOut() #sklearn way
+        weights = self.weights
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+           
+            #######There may need some hints how to use it better 
+            ### also if we need to add weight option here
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom += awkde.evaluate(test_data)  
+        return np.mean(fom)
+
+    def kfold_cv_score(self, bandwidth_val, alpha_val, n_splits=2):
+        """
+        to do: how to choose and use s_split in init
+        """
+        kf = KFold(n_splits=n_splits, shuffle=True, random_state=42)
+        fom = []
+        for train_index, test_index in kf.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom.append(awkde.evaluate(test_data).sum())
+        return sum(fom)
+
+    def optimize_parameters(self, method='loo_cv', fom_plot=False):
+        best_score = float('inf')
+        best_params = {'bandwidth': None, 'alpha': None}
+
+        FOM= {}
+        for bandwidth in self.bandwidth_options:
+            for alpha in self.alpha_options:
+                if method=='kfold_cv':
+                    score = self.kfold_cv_score(bandwidth, alpha, n_splits=2)
+                else:
+                    score = self.loo_cv_score(bandwidth, alpha)
+
+                FOM[(bandwidth, alpha)] = score
+                if score < best_score:
+                    best_score = score
+                    best_params['bandwidth'] = bandwidth
+                    best_params['alpha'] = alpha
+        
+        if fom_plot==True:
+            import operator
```

Comment:
since operator is a standard python library, I think it's OK to import it at the start of the module

Reply:
done